### DISCUSSION TITLE
Lights scene went too dark in my program.

### DISCUSSION THREAD

**Discussion by Bowen951209** [RayTracing/raytracing.github.io#1653] [Q&A] [answered]
Hello. I'm working on emissive materials in book 2, chapter 7. I got some similar results as the book's from the simple light scene and the Cornell box scene, but the result is very dark.

### My simple light scene:
![image](https://github.com/user-attachments/assets/9460422f-e506-47cc-9f3f-19ccc90cf9eb)

### My Cornell box:
![image](https://github.com/user-attachments/assets/a335e13b-d33f-4529-a66e-fb43bef94cb8)
(The back wall is gone, which is also a bug I'm facing.)

Both results were rendered at 100 samples and max ray depth of 50.

Because I wrote the code in GLSL, I tried to use iteration rather than recursion to trace rays. I don't know if my implementation was right, but I do get correct results from the previous scenes that didn't include emissive materials.

```GLSL
vec3 get_color(Ray ray) {
    vec3 final_color = vec3(0.0);
    vec3 accumulated_attenuation = vec3(1.0); // Start with no attenuation

    HitRecord hit_record;
    // Loop until we either reach the maximum recursion depth or stop scattering.
    for (int i = 0; i < max_depth; i++) {
        if (!trace_through_bvh(ray, Interval(0.001, INFINITY), hit_record)) {
            final_color = accumulated_attenuation * background;
            break;
        }

        if(!scatter(ray.dir, hit_record.normal, hit_record.is_front_face, material)) {
            final_color = accumulated_attenuation * color_from_emission;
            break;
        }

        // Update ray origin.
        ray.o = hit_record.p;

        accumulated_attenuation *= albedo;
    }

    return linear_to_gamma(final_color);
}
```
In the code, `ray.dir` is updated by the `scatter()` function, and `color_from_emission` and `albedo` are global variables that will be updated if an object is hit.

I'm confused because rays actually had a low chance to reach the light source, which would in turn produce a dark scene just like mine. So I suppose that my understanding may somehow be wrong. 

My repository is available [here](https://github.com/Bowen951209/raytracing-book/tree/3203ee335a44186ca21db6d50b3ab0a51b9fa405).

**Comment by hollasch** [comment]
You might find it helpful to compare your debug flow against that of our "progression" branches, where we commit increments along with the code progression in the books. If you look at branch "progression/2024-0405" (our latest progression), you'll find commit 68f312d, which is our reference code immediately after book 2 listing 60. Other readers have found this to be a very helpful debug tool.

**Reply by Bowen951209** [reply]
Thanks, I'll check that. But I don't just quite understand why rays that seldom reach the light source still produce a bright result in the book.

**Comment by Bowen951209** [comment, marked as answer]
I found the problem! It’s related to gamma correction. I think there’s a bug in my gamma correction GLSL code. To resolve it, I used the OpenGL built-in method to enable gamma correction, which applies a gamma value of 2.2.

![image](https://github.com/user-attachments/assets/7ac7b090-34b1-4b24-a02b-ceb5761f0227)

**Reply by Bowen951209** [reply]
By the way, the missing wall is due to precision loss in the padding float when it’s passed to the GPU.
